src/tds/convert.c

```diff
@@ -1538,13 +1538,13 @@ tds_convert_real(const TDS_REAL* src, int desttype, CONV_RESULT * cr)
 		return sizeof(TDS_UINT);
 		break;
 	case SYBINT8:
-		if (the_value > (TDS_REAL) TDS_INT8_MAX || the_value < (TDS_REAL) TDS_INT8_MIN)
+		if (!IS_INT8(the_value))
 			return TDS_CONVERT_OVERFLOW;
 		cr->bi = (TDS_INT8) the_value;
 		return sizeof(TDS_INT8);
 		break;
 	case SYBUINT8:
-		if (the_value > (TDS_REAL) TDS_UINT8_MAX || the_value < 0)
+		if (the_value >= TDS_UINT8_MAX + 1.0f || the_value < 0)
```

Comment:
Wondering if this `TDS_UINT8_MAX + 1.0f` is portable or not. Technically the compiler will convert the first number to float then do the addition. However nor float nor double have the precision to hold all the number precisely (double has 54 bit precision while he number is 63 bit precise). It will probably round directly the number, then add 1 (which won't change). Potentially the result is what we are expecting, but hard to say. If the compile would do truncation on the first conversion the result is not the expected number.

Reply:
Why not having something like
```c
#define TDS_UINT8_UPPER 18446744073709551616.0
```

Reply:
From what I gather (details below), the `TDS_UINT8_MAX` approach should be pretty portable.  Upon consideration, though, a third possibility could be to take advantage of the fact that `TDS_UINT8_MIN` should have an exact representation by multiplying it by `-2.0` -- per https://stackoverflow.com/questions/29174165/is-flt-radix-ever-not-2-in-c11-for-modern-general-purpose-computers, in the rare event that `FLT_RADIX` isn't two, it should still be a power of two, which suffices (with decimal floating-point, if supported, having separate `_DecimalN` types).

As for rounding, per https://en.cppreference.com/w/c/language/conversion#Real_floating-integer_conversions, on implementations supporting IEEE floating point (which should be more or less all of them), rounding is to nearest for anything evaluable at compile time unless `FENV_ACCESS` (https://en.cppreference.com/w/c/preprocessor/impl) is `ON` (unusual) *and* the application has used `fesetround` to specify some other rounding mode.

Meanwhile, I'd forgotten that the first step in conversion from floating point was truncation rather than taking a floor, so I'll want to allow more margin on signed lower bounds accordingly.